### PR TITLE
[SPARK-38159][SQL] Add a new FileSourceMetadataAttribute for the Hidden File Metadata

### PR DESCRIPTION
**Pull request by Yaohua628**
### What changes were proposed in this pull request?
Add a new `FileSourceMetadataAttribute` object with an `apply` method to create a `FileSourceMetadataAttribute`, and an `unapply` method to match only file source metadata attribute.

### Why are the changes needed?
Extra safeguard to make sure it matches file source metadata attribute.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Existing UTs

### CODE REVIEW

**Comment by Yaohua628** [comment]
Hi @cloud-fan, an XS PR to refactor the `MetadataAttribute` `unapply` method, please take a look whenever you have time, thanks a lot!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala [hunk 1]
```diff
@@ -437,16 +437,20 @@ object VirtualColumn {
  * The internal representation of the hidden metadata struct:
  * set `__metadata_col` to `true` in AttributeReference metadata
  * - apply() will create a metadata attribute reference
- * - unapply() will check if an attribute reference is the metadata attribute reference
+ * - unapply() will check if an attribute reference is the file source metadata attribute reference
  */
 object MetadataAttribute {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala:442]: Maybe we should have two objects
1. `MetadataAttribute`: This is used to create and extract general metadata attributes (not only for file source)
2. `FileSourceMetadataAttribute`: To extract file source metadata attributes

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala:442]: Or we make sure that we only call `MetadataAttribute.unapply` in places that only deal with file sources.

Comment (Yaohua628) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala:442]: make sense!

**Comment by Yaohua628** [comment]
@cloud-fan addressed your comment by adding a `FileSourceMetadataAttribute`, thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala [hunk 2]
```diff
@@ -451,3 +451,21 @@ object MetadataAttribute {
...
+ * The internal representation of the FileSourceMetadataAttribute,
+ * it sets `__metadata_col` to `true` in AttributeReference metadata with the attr name `_metadata`
+ * - unapply() will check if an attribute reference is the file source metadata attribute reference
+ */
+object FileSourceMetadataAttribute {
+
+  val ATTRIBUTE_NAME = "_metadata"
+
+  def unapply(attr: AttributeReference): Option[AttributeReference] = {
+    if (attr.metadata.contains(METADATA_COL_ATTR_KEY)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala:465]: nit:
```
attr match {
  case MetadataAttribute(a) if a.name.equalsIgnoreCase(ATTRIBUTE_NAME) => Some(attr)
  case _ => None
}
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala [hunk 3]
```diff
@@ -214,12 +214,12 @@ object FileSourceStrategy extends Strategy with PredicateHelper with Logging {
       logInfo(s"Output Data Schema: ${outputSchema.simpleString(5)}")
 
       val metadataStructOpt = l.output.collectFirst {
-        case MetadataAttribute(attr) => attr
+        case FileSourceMetadataAttribute(attr) => attr
       }
 
       val metadataColumns = metadataStructOpt.map { metadataStruct =>
         metadataStruct.dataType.asInstanceOf[StructType].fields.map { field =>
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala:221]: now we can do `metadataStruct.dataType.asInstanceOf[StructType].toAttributes`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala [hunk 4]
```diff
@@ -366,9 +364,8 @@ case class FileSourceScanExec(
   @transient
   private lazy val pushedDownFilters = {
     val supportNestedPredicatePushdown = DataSourceUtils.supportNestedPredicatePushdown(relation)
-    // TODO: should be able to push filters containing metadata columns down to skip files
     dataFilters.filterNot(_.references.exists {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala:367]: I think it's better to remove the metadata col filter in `FileSourceStrategy`, as that's where we make the metadata col filter invalid: we remove the struct-type metadata col and convert its fields to attributes.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala [hunk 5]
```diff
@@ -451,3 +451,21 @@ object MetadataAttribute {
...
 }
+
+/**
+ * The internal representation of the FileSourceMetadataAttribute,
+ * it sets `__metadata_col` to `true` in AttributeReference metadata with the attr name `_metadata`
+ * - unapply() will check if an attribute reference is the file source metadata attribute reference
+ */
+object FileSourceMetadataAttribute {
+
+  val ATTRIBUTE_NAME = "_metadata"
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala:462]: After more thought, I think it's more robust to not rely on the attribute name to extract file source metadata attrs. We can add a new metadata key to indicate file source metadata attributes. Sorry for the back and forth!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala (hunk 4 above)
Comment (Yaohua628) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala:370]: agreed, it is better to remove in `FileSourceStrategy`, but we actually need to use the intact `dataFilters` here in `FileSourceScanExec` to filter files to read (i.e. `listFiles(..., dataFilters)`). 

added some comments here, do you think it is OK for now?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala (hunk 5 above)
Comment (Yaohua628) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala]: np! sounds good to me!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala [hunk 6]
```diff
@@ -451,3 +451,28 @@ object MetadataAttribute {
...
   }
 }
+
+/**
+ * The internal representation of the FileSourceMetadataAttribute, it sets `__metadata_col`
+ * and `__file_source_metadata_col` to `true` in AttributeReference's metadata
+ * - apply() will create a file source metadata attribute reference
+ * - unapply() will check if an attribute reference is the file source metadata attribute reference
+ */
+object FileSourceMetadataAttribute {
```
Comment (Yaohua628) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala:461]: still keep the `METADATA_COL_ATTR_KEY` in `FileSourceMetadataAttribute`: `FileSourceMetadataAttribute` <= `MetadataAttribute`

**Comment by Yaohua628** [comment]
@cloud-fan addressed all your comments: 
- added a new key for `FileSourceMetadataAttribute`
- addressed some nits
- replied one of your suggestions

ready for another pass, thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala [hunk 7]
```diff
@@ -196,12 +196,10 @@ case class FileSourceScanExec(
     optionalNumCoalescedBuckets: Option[Int],
     dataFilters: Seq[Expression],
     tableIdentifier: Option[TableIdentifier],
-    disableBucketedScan: Boolean = false)
+    disableBucketedScan: Boolean = false,
+    metadataColumns: Seq[AttributeReference])
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala:200]: Now we can avoid adding this new parameter. When convert the fields of struct-type metadata col, we can use `FileSourceMetadataAttribute`

Comment (Yaohua628) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala:200]: oops, forgot to revert it back...

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
document building fails. Can you rerun the Github Action job?

**Comment by Yaohua628** [comment]
> document building fails. Can you rerun the Github Action job?

done! thanks

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**